### ISSUE TITLE
Bogus RS0026 // Do not add multiple public overloads with optional parameters

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#16564]
The Roslyn analyzer reports bogus errors RS0026. 

For example for completely sound public APIs:

```C#
public static Script<T> Create<T>(string code, ScriptOptions options = null, Type globalsType = null, InteractiveAssemblyLoader assemblyLoader = null)
public static Script<T> Create<T>(Stream code, ScriptOptions options = null, Type globalsType = null, InteractiveAssemblyLoader assemblyLoader = null)
```

We already have 10 suppressions of this in the code base. Seems like the analyzer should be fixed to not flag valid overloads.

**Comment by tmat** [comment]
@filipw FYI

**Comment by tmat** [comment]
@jaredpar

**Comment by mavasani** [comment]
This analyzer diagnostic was added to enforce: https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md

See section `DO NOT add multiple overloads with optional parameters`. Also tagging @agocke @jaredpar, who have contributed to this document.

**Comment by kzu** [comment]
Another case where this analyzer is bogus IMO:

```csharp
Task ExecuteAsync(IAsyncCommand command, CancellationToken cancellation = default);
Task<TResult> ExecuteAsync<TResult>(IAsyncCommand<TResult> command, CancellationToken cancellation = default);
```

The analyzer things both are overloads of each other, where one is generic and the other isn't. I wouldn't expect this case to be flagged.
